File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/forth/forth.go
package forth

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

type stack struct {
	data []int
}

func (s *stack) push(v int) {
	s.data = append(s.data, v)
}

func (s *stack) pop() (int, error) {
	if len(s.data) == 0 {
		return 0, errors.New("stack underflow")
	}
	v := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	return v, nil
}

func (s *stack) peek() (int, error) {
	if len(s.data) == 0 {
		return 0, errors.New("insufficient elements")
	}
	return s.data[len(s.data)-1], nil
}

func (s *stack) dup() error {
	v, err := s.peek()
	if err != nil {
		return err
	}
	s.push(v)
	return nil
}

func (s *stack) swap() error {
	if len(s.data) < 2 {
		return errors.New("insufficient elements")
	}
	a, _ := s.pop()
	b, _ := s.pop()
	s.push(a)
	s.push(b)
	return nil
}

func (s *stack) over() error {
	if len(s.data) < 2 {
		return errors.New("insufficient elements")
	}
	v := s.data[len(s.data)-2]
	s.push(v)
	return nil
}

var builtins = map[string]func(*stack) error{
	"+": func(s *stack) error {
		a, err := s.pop()
		if err != nil {
			return err
		}
		b, err := s.pop()
		if err != nil {
			return err
		}
		s.push(b + a)
		return nil
	},
	"-": func(s *stack) error {
		a, err := s.pop()
		if err != nil {
			return err
		}
		b, err := s.pop()
		if err != nil {
			return err
		}
		s.push(b - a)
		return nil
	},
	"*": func(s *stack) error {
		a, err := s.pop()
		if err != nil {
			return err
		}
		b, err := s.pop()
		if err != nil {
			return err
		}
		s.push(b * a)
		return nil
	},
	"/": func(s *stack) error {
		a, err := s.pop()
		if err != nil {
			return err
		}
		if a == 0 {
			return errors.New("division by zero")
		}
		b, err := s.pop()
		if err != nil {
			return err
		}
		s.push(b / a)
		return nil
	},
	"DUP":  func(s *stack) error { return s.dup() },
	"DROP": func(s *stack) error { _, err := s.pop(); return err },
	"SWAP": func(s *stack) error { return s.swap() },
	"OVER": func(s *stack) error { return s.over() },
}

func Forth(input []string) ([]int, error) {
	s := &stack{}
	customWords := make(map[string][]string)

	for _, line := range input {
		tokens := tokenize(line)
		if len(tokens) == 0 {
			continue
		}

		// Expand custom words and process tokens
		processed := []string{}
		for i := 0; i < len(tokens); i++ {
			token := tokens[i]
			upperToken := strings.ToUpper(token)

			if upperToken == ":" {
				// Handle word definition
				if i+3 > len(tokens) {
					return nil, errors.New("invalid word definition")
				}
				wordName := strings.ToUpper(tokens[i+1])
				if _, err := strconv.Atoi(wordName); err == nil {
					return nil, fmt.Errorf("invalid word name: %s", wordName)
				}
				end := i + 2
				for ; end < len(tokens) && tokens[end] != ";"; end++ {
				}
				if end >= len(tokens) {
					return nil, errors.New("unterminated word definition")
				}
				processedDef := make([]string, 0, len(tokens[i+2:end]))
				for _, t := range tokens[i+2 : end] {
					if _, err := strconv.Atoi(t); err == nil {
						processedDef = append(processedDef, t)
					} else {
						processedDef = append(processedDef, strings.ToUpper(t))
					}
				}
				customWords[wordName] = processedDef
				i = end // skip to end of definition
				continue
			}

			if def, ok := customWords[upperToken]; ok {
				processed = append(processed, def...)
			} else {
				processed = append(processed, upperToken)
			}
		}

		// Evaluate processed tokens
		for _, token := range processed {
			if num, err := strconv.Atoi(token); err == nil {
				s.push(num)
				continue
			}

			op, ok := builtins[token]
			if !ok {
				return nil, fmt.Errorf("undefined word: %s", token)
			}

			if err := op(s); err != nil {
				return nil, err
			}
		}
	}

	return s.data, nil
}

func tokenize(line string) []string {
	var tokens []string
	var current strings.Builder

	for _, r := range line {
		if r == ' ' {
			if current.Len() > 0 {
				tokens = append(tokens, current.String())
				current.Reset()
			}
		} else {
			current.WriteRune(r)
		}
	}
	if current.Len() > 0 {
		tokens = append(tokens, current.String())
	}

	return tokens
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/python/exercises/practice/forth/forth.py
class StackUnderflowError(Exception):
    """Exception raised when Stack is not full.
       message: explanation of the error.
    """
    def __init__(self, message):
        self.message = message


def evaluate(input_data):
    """Evaluate Forth code and return the stack state."""
    from collections import deque

    custom_words = {}
    stack = []
    
    for line in input_data:
        tokens = line.split()
        if not tokens:
            continue
        
        # Handle word definitions
        if tokens[0] == ":" and tokens[-1] == ";":
            if len(tokens) < 4:
                raise ValueError("invalid word definition")
            name = tokens[1].upper()
            # Check if trying to redefine a number
            try:
                int(name)
                raise ValueError("cannot redefine numbers")
            except ValueError:
                pass  # Name is not a number
            
            # Process definition tokens (convert words to uppercase, leave numbers)
            definition = []
            for token in tokens[2:-1]:
                try:
                    int(token)
                    definition.append(token)
                except ValueError:
                    definition.append(token.upper())
            custom_words[name] = definition
            
        else:
            # Process normal line of code
            processing_queue = deque()
            for token in tokens:
                # Separate numbers and words (convert words to uppercase)
                try:
                    int(token)
                    processing_queue.append(token)
                except ValueError:
                    processing_queue.append(token.upper())
            
            while processing_queue:
                token = processing_queue.popleft()
                
                try:
                    # Try to push number
                    num = int(token)
                    stack.append(num)
                except ValueError:
                    # Handle word/operation
                    word = token
                    if word in custom_words:
                        # Convert definition tokens to appropriate format and prepend
                        def_tokens = []
                        for def_token in custom_words[word]:
                            try:
                                # Check if it's a number (preserve as string)
                                int(def_token)
                                def_tokens.append(def_token)
                            except ValueError:
                                # Convert word to uppercase
                                def_tokens.append(def_token.upper())
                        # Prepend in reverse order to maintain execution order
                        for def_token in reversed(def_tokens):
                            processing_queue.appendleft(def_token)
                    else:
                        # Built-in operations
                        if word == "+":
                            if len(stack) < 2:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            a = stack.pop()
                            b = stack.pop()
                            stack.append(b + a)
                        elif word == "-":
                            if len(stack) < 2:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            a = stack.pop()
                            b = stack.pop()
                            stack.append(b - a)
                        elif word == "*":
                            if len(stack) < 2:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            a = stack.pop()
                            b = stack.pop()
                            stack.append(b * a)
                        elif word == "/":
                            if len(stack) < 2:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            a = stack.pop()
                            b = stack.pop()
                            if a == 0:
                                raise ZeroDivisionError("divide by zero")
                            # Forth uses integer division truncating toward zero
                            stack.append(b // a)
                        elif word == "DUP":
                            if len(stack) < 1:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            stack.append(stack[-1])
                        elif word == "DROP":
                            if len(stack) < 1:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            stack.pop()
                        elif word == "SWAP":
                            if len(stack) < 2:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            a = stack.pop()
                            b = stack.pop()
                            stack.append(a)
                            stack.append(b)
                        elif word == "OVER":
                            if len(stack) < 2:
                                raise StackUnderflowError("Insufficient number of items in stack")
                            stack.append(stack[-2])
                        else:
                            raise ValueError(f"undefined operation {word}")
    
    return stack
